### PR TITLE
[ESP32] CLI option to set delayed action time in OTA Provider app and restart OTA Requestor once ApplyUpdate is successful

### PR DESCRIPTION
**Pull request by shubhamdp**
#### Problem
* No option to set delayed action time in ESP32 OTA Provider application
* We had to manually restart the ESP32 device once the ApplyUpdate is successful

#### Change overview
* Added an cli option to set delayed action time in ESP32 OTA Provider application
* Modified the Linux OTA Provider help for delayed action time cli option
* Schedule the restart for ESP32 once apply update is successful
* Also, Apply suggestion from #13484 in lighting-app
* Minor changes in readme files

#### Testing
* Linux OTA provider and ESP32 OTA requestor app
* ESP32 OTA provider and tried OTA in ESP32 lighting app

### CODE REVIEW

File: examples/ota-provider-app/esp32/main/OTAProviderCommands.cpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    // Register subcommands of the `OTAProvider` commands.
+    static const shell_command_t subCommands[] = {
+        { &DelayedActionTimeHandler, "delay",
+          "Set delayed action time for QueryImageResponse and ApplyUpdateResponse\n"
+          "Usage: OTAProvider delay <delay in seconds>" },
+    };
+
+    sSubShell.RegisterCommands(subCommands, ArraySize(subCommands));
+
+    // Register the root `ota` command in the top-level shell.
```
Comment (dhrishi) [inline examples/ota-provider-app/esp32/main/OTAProviderCommands.cpp:74]: Nit:OTA provider?

File: examples/lighting-app/esp32/README.md [hunk 2]
```diff
@@ -154,10 +154,17 @@ After commissioning is successful, press Enter in requestor device console and
...
 
 ```
->matter ota query 1 12345 0
+>matter ota query 1 12346 0
+```
+
+Alternate way to query an OTA image using chip-tool
+
+```
+./out/debug/chip-tool otasoftwareupdaterequestor announce-ota-provider 12346 0 0 0 12345 0
```
Comment (dhrishi) [inline examples/lighting-app/esp32/README.md:163]: Please see if you want to remove this to avoid confusion, as per our discussion offline

**Comment by dhrishi** [comment]
@shubhamdp Please resolve the merge conflicts

File: examples/lighting-app/esp32/README.md (hunk 2 above)
Comment (shubhamdp) [reply examples/lighting-app/esp32/README.md]: Removed this option.

File: examples/lighting-app/esp32/README.md [hunk 3]
```diff
@@ -154,10 +154,11 @@ After commissioning is successful, press Enter in requestor device console and
...
 ```
->matter ota query 1 12345 0
+>matter ota query 1 12346 0
 ```
 
 ## Apply update
 
-Once transfer is complete, reboot the device manually to boot from upgraded OTA
-image.
+Once the transfer is complete OTA requestor sends ApplyUpdateRequest command to
```
Comment (carol-apple) [inline examples/lighting-app/esp32/README.md:162]: nit
```suggestion
Once the transfer is complete, OTA requestor sends ApplyUpdateRequest command to
```

File: examples/ota-requestor-app/esp32/README.md [hunk 4]
```diff
@@ -45,8 +45,9 @@ chip-tool. On receiving this command OTA requestor will query for OTA image.
 
 ## Apply update
 
-Once transfer is complete, reboot the device manually to boot from upgraded OTA
-image.
+Once the transfer is complete OTA requestor sends ApplyUpdateRequest command to
```
Comment (carol-apple) [inline examples/ota-requestor-app/esp32/README.md:48]:
```suggestion
Once the transfer is complete, OTA requestor sends ApplyUpdateRequest command to
```

File: src/platform/ESP32/OTAImageProcessorImpl.cpp [hunk 5]
```diff
@@ -26,9 +26,17 @@
...
 #define TAG "OTAImageProcessor"
+using namespace chip::System;
 using namespace ::chip::DeviceLayer::Internal;
 
 namespace chip {
+namespace {
+
+void HandleRestart(Layer * systemLayer, void * appState)
+{
+    esp_restart();
```
Comment (carol-apple) [inline src/platform/ESP32/OTAImageProcessorImpl.cpp:37]: After restart, NotifyUpdateApplied should also be called once the device has booted to a new version. Is that something you intend on doing in a later PR?

Comment (shubhamdp) [reply src/platform/ESP32/OTAImageProcessorImpl.cpp:37]: Yes, I'll add a follow-up PR for that.

File: examples/ota-provider-app/esp32/main/OTAProviderCommands.cpp [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    return error;
+}
+} // namespace
+
+void OTAProviderCommands::Register()
+{
+    // These commands can be moved to src/lib/shell/commands/Ota.cpp along with the other OTA commands.
+    // But as of now only Linux and ESP32 platforms supprts OTA provider
```
Comment (dhrishi) [inline examples/ota-provider-app/esp32/main/OTAProviderCommands.cpp:63]: Typo: support

**Review by dhrishi — APPROVED** [review]

File: src/platform/ESP32/OTAImageProcessorImpl.cpp (hunk 5 above)
Comment (shubhamdp) [reply src/platform/ESP32/OTAImageProcessorImpl.cpp:37]: Added an issue for the same, #13760

File: examples/lighting-app/esp32/README.md (hunk 3 above)
Comment (carol-apple) [reply examples/lighting-app/esp32/README.md:162]: I believe this comma is still missing?

File: examples/ota-requestor-app/esp32/README.md (hunk 4 above)
Comment (carol-apple) [reply examples/ota-requestor-app/esp32/README.md:48]: Same here

**Review by carol-apple — APPROVED** [review]

File: examples/lighting-app/esp32/README.md (hunk 3 above)
Comment (shubhamdp) [reply examples/lighting-app/esp32/README.md]: I think I added the suggestions to batch and forgot to submit. Submitted now.

**Review by holbrookt — APPROVED** [review]

**Pull request merged**